File: 10-GRAPHS/Python/course-schedule.py
from typing import List

# https://leetcode.com/problems/course-schedule/

class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        graph = {i: [] for i in range(numCourses)}
        for course, prereq in prerequisites:
            graph[course].append(prereq)

        visited = [False] * numCourses

        def dfs(node: int) -> bool:
            if visited[node]:
                return False
            if not graph[node]:
                return True

            visited[node] = True

            for neighbor in graph[node]:
                if not dfs(neighbor):
                    return False

            visited[node] = False
            graph[node] = []
            return True

        for i in range(numCourses):
            if not dfs(i):
                return False

        return True

File: 10-GRAPHS/Go/course-schedule.go
package main

// https://leetcode.com/problems/course-schedule/

func canFinish(numCourses int, prerequisites [][]int) bool {
	graph := make(map[int][]int, numCourses*2)
	for _, pair := range prerequisites {
		course := pair[0]
		graph[course] = append(graph[course], pair[1])
	}

	visited := make([]bool, numCourses)

	var dfs func(node int) bool
	dfs = func(node int) bool {
		if visited[node] {
			return false
		}
		if len(graph[node]) == 0 {
			return true
		}

		visited[node] = true

		for _, neighbour := range graph[node] {
			if !dfs(neighbour) {
				return false
			}
		}

		visited[node] = false
		graph[node] = []int{}
		return true
	}

	for i := 0; i < numCourses; i++ {
		if !dfs(i) {
			return false
		}
	}

	return true
}
